File: 이진검색(이진탐색)/leetcode2300(Medium)-Successful Pairs of Spells and Potions.py
# 1681ms(64.59%), 39.5MB(11.5%)
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        dp = {}
        potions.sort()

        def search(co):
            left, right = 0, len(potions) - 1
            while left < right:
                mid = left + ((right - left) // 2)
                num = potions[mid] * co
                if num < success:
                    left = mid + 1
                else:
                    right = mid
            return left

        ans = []
        for spell in spells:
            if spell not in dp:
                index = search(spell)
                pairs_count = 0 if potions[index] * spell < success else len(potions) - index
                ans.append(pairs_count)
                dp[spell] = pairs_count
                continue
            ans.append(dp[spell])
        return ans

File: 이진검색(이진탐색)/leetcode2300(Medium)-Successful Pairs of Spells and Potions.js
// 339ms(35.56%), 67MB(97%)
/**
 * @param {number[]} spells
 * @param {number[]} potions
 * @param {number} success
 * @return {number[]}
 */
var successfulPairs = function (spells, potions, success) {
  const dp = {}
  potions = potions.sort((a, b) => a - b)

  const search = (co) => {
    let left = 0
    let right = potions.length - 1
    while (left < right) {
      const mid = left + Math.floor((right - left) / 2)
      const num = potions[mid] * co
      if (num < success) {
        left = mid + 1
      } else {
        right = mid
      }
    }
    return left
  }

  return spells.map((spell) => {
    if (!dp[spell]) {
      const index = search(spell)
      const pairsCount =
        potions[index] * spell < success ? 0 : potions.length - index
      dp[spell] = pairsCount
    }
    return dp[spell]
  })
}
